google-cloud-firestore/src/main/java/com/google/cloud/firestore/telemetry/EnabledMetricsUtil.java

```diff
@@ -0,0 +1,232 @@
+    Set<String> attributesFilter =
+        ImmutableSet.<String>builder()
+            .addAll(
+                COMMON_ATTRIBUTES.stream().map(AttributeKey::getKey).collect(Collectors.toSet()))
+            .build();
+    View view = View.builder().setAttributeFilter(attributesFilter).build();
+
+    viewMap.put(selector, view);
+  }
+
+  private void addTracerFactory(
+      List<ApiTracerFactory> apiTracerFactories, BuiltinMetricsProvider metricsProvider) {
+    ApiTracerFactory tracerFactory = metricsProvider.getApiTracerFactory();
+    if (tracerFactory != null) {
+      apiTracerFactories.add(tracerFactory);
+    }
+  }
+
+  class MetricsContext implements MetricsUtil.MetricsContext {
+    private final Stopwatch stopwatch;
+    private final String methodName;
+
+    public MetricsContext(String methodName) {
+      this.stopwatch = Stopwatch.createStarted();
+      this.methodName = methodName;
+    }
+
+    public <T> void recordEndToEndLatencyAtFuture(ApiFuture<T> futureValue) {
+      ApiFutures.addCallback(
+          futureValue,
+          new ApiFutureCallback<T>() {
+            @Override
+            public void onFailure(Throwable t) {
+              recordEndToEndLatency(t);
+            }
+
+            @Override
+            public void onSuccess(T result) {
+              recordEndToEndLatency();
+            }
+          },
+          MoreExecutors.directExecutor());
+    }
+
+    public void recordEndToEndLatency() {
+      recordEndToEndLatency(StatusCode.Code.OK.toString());
+    }
+
+    public void recordEndToEndLatency(Throwable t) {
+      recordEndToEndLatency(extractErrorStatus(t));
+    }
+
+    public void recordFirstResponseLatency() {
+      double elapsedTime = stopwatch.elapsed(TimeUnit.MILLISECONDS);
+      Map<String, String> attributes = createAttributes(StatusCode.Code.OK.toString());
+      defaultMetricsProvider.firstResponseLatencyRecorder(elapsedTime, attributes);
+      customMetricsProvider.firstResponseLatencyRecorder(elapsedTime, attributes);
+    }
+
+    private void recordEndToEndLatency(String status) {
+      double elapsedTime = stopwatch.elapsed(TimeUnit.MILLISECONDS);
+      Map<String, String> attributes = createAttributes(status);
+      defaultMetricsProvider.endToEndRequestLatencyRecorder(elapsedTime, attributes);
+      customMetricsProvider.endToEndRequestLatencyRecorder(elapsedTime, attributes);
+    }
+
+    private Map<String, String> createAttributes(String status) {
+      Map<String, String> attributes = new HashMap<>();
+      attributes.put(METRIC_KEY_METHOD.getKey(), methodName);
+      attributes.put(METRIC_KEY_STATUS.getKey(), status);
+      return attributes;
+    }
+
+    private String extractErrorStatus(@Nullable Throwable throwable) {
+      if (!(throwable instanceof FirestoreException)) {
+        return StatusCode.Code.UNKNOWN.toString();
+      }
+      Status status = ((FirestoreException) throwable).getStatus();
```

Comment:
`FirestoreException::getStatus()` may return `null`. So the code should check for that

```java
Status status = ((FirestoreException) throwable).getStatus();
if (status == null) {
  return StatusCode.Code.UNKNOWN.toString();
}
return status.getCode().name();
```

Reply:
turns out, at least in transactions, getStatus() always return null.